### PR TITLE
add array of headers as optional parameter

### LINKED ISSUE #265: feature: Artifactory authentication support
Currently archive::artifactory doen't not accept username and password.
It will be very helpful if we can download artifacts with authentication

### LINKED ISSUE #445: Unable to download from an artifactory host with X-JFrog-Art-Api Header
We have to use a protected artifactory to download a package, the only option is to use X-JFrog-Art-Api Header, user+pass would be 2FA protected, so will not work.

- Puppet: 7.1.2
- Ruby: 
- Distribution: ubuntu 
- Module version: master 

## How to reproduce (e.g Puppet code you use)
Try to use download_options with a X-JFrog-Art-Api Header (which I understood should be usable for that ?)

```puppet
archive { "package.tar.gz":
    ensure           => present,
    path             => $target_file,
    source           => "${artifactory_url}/${repo}/${branch}/package.tar.gz",
    checksum_type    => 'md5',
    checksum_verify  => $package_md5,
    checksum_url     => "$artifactory_url}/${repo}/${branch}/package.tar.gz.md5",
    download_options => "-H \"X-JFrog-Art-Api: ${token}\"",
}
```

## What are you seeing
`Error: /Stage[main]/Application::Deployer/Archive[package.tar.gz]/ensure: change from 'absent' to 'present' failed: Execution of '/usr/bin/curl https://artifactory/artifactory/.../...../package.tar.gz -o /srv/tmp_deployment/package.tar.gz_20210617-72402-a7acae -fsSLg --max-redirs 5 -H "X-JFrog-Art-Api: XXXXXXXXXXXXX"' returned 22: curl: (22) The requested URL returned error: 401 Unauthorized`

## What behaviour did you expect instead
Working

## Output log
see above

## Any additional information you'd like to impart
When copying the whole command line to bash as-is, downloading works (??)

### LINKED ISSUE #476: add support for Ubuntu 22.04
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: all currently supported
- Ruby: all currently supported
- Distribution: all currently supported
- Module version: next major

## How to reproduce (e.g Puppet code you use)

install module

## What are you seeing

it works, extended acceptance tests

## What behaviour did you expect instead

n/a - already fixed and passing, existing PR will close this issue

## Output log

n/a

## Any additional information you'd like to impart

thank you for letting me participate

### PR DESCRIPTION
**Pull request by prolixalias**
linting

update in-line puppet strings documentation

added tests for passing headers

update README for headers

add Ubuntu 22.04 to metadata

rebase 'modulesync 5.3.0' work 

#### Pull Request (PR) description

add headers param, update tests and documentation (strings, readme)

incorporate modulesync #463

#### This Pull Request (PR) fixes the following issues

Fixes #265
Fixes #445
Fixes #476
Finishes #463

### CODE REVIEW

File: manifests/artifactory.pp [hunk 1]
```diff
@@ -1,58 +1,86 @@
...
-# Parameters
-# ----------
+# @param [Stdlib::HTTPUrl] url
+#   artifactory download URL
+# @param [Array] headers
+#   HTTP header(s) to pass to source
+#   Default value: []
+# @param [String] path
+#   absolute path for the download file (or use archive_path and only supply filename)
+#   Default value: namevar
```
Comment (kenyon) [inline manifests/artifactory.pp:13]: remove defaults and data types. puppet-strings documents those automatically.

File: spec/unit/puppet/provider/archive/curl_spec.rb [hunk 2]
```diff
@@ -15,7 +15,8 @@
       [
         'http://home.lan/example.zip',
         '-o',
-        String,
+        # String,
```
Comment (kenyon) [inline spec/unit/puppet/provider/archive/curl_spec.rb:18]: looks like this can just be removed, not sure how this was working before

**Review by kenyon — CHANGES_REQUESTED** [review]

File: manifests/artifactory.pp (hunk 1 above)
Comment (prolixalias) [reply manifests/artifactory.pp]: fixed, thanks for the help!

File: spec/unit/puppet/provider/archive/curl_spec.rb (hunk 2 above)
Comment (prolixalias) [reply spec/unit/puppet/provider/archive/curl_spec.rb]: done.

**Review by kenyon — APPROVED** [review]

**Comment by prolixalias** [comment]
@bastelfreak & @kenyon - was able to finish (I believe) #463 and re-request review. However, the acceptance test on archlinuxrolling seems to be a problem. I'm happy to assist on that one too. However, it looks like that's in another repo. Should I take a stab at it or is there more to it (that failure) than meets the eye?

**Comment by bastelfreak** [comment]
Ah here you fixed the escape problem. I guess the issue on arch linux is that wget isn't installed by default in the docker image. you could do something like `package { 'wget': ensure => 'installed' }` in the acceptancetest to ensure it's always installed.

**Review by prolixalias — COMMENTED** [review]
looks like it's ready to merge

**Comment by prolixalias** [comment]
Is there anything I need to do to help this one get bumped up in priority? My current engagement has some unusual requirements about pointing to things that aren't tags from forge... They keep asking me for updates and I don't have any. :)

**Review by bastelfreak — APPROVED** [review]

**Comment by bastelfreak** [comment]
thanks for all the work!

**Pull request merged**